backend/tests/api_e2e/test_api_allocations_ws.py

```diff
@@ -0,0 +1,328 @@
+
+    # Store results in this object
+    results = SocketIOTestResults()
+
+    async def run_test():
+        # Capture server logs
+        server_log = io.StringIO()
+        log_handler = logging.StreamHandler(server_log)
+        log_handler.setLevel(logging.WARNING)
+        uvicorn_logger = logging.getLogger("uvicorn")
+        uvicorn_logger.addHandler(log_handler)
+
+        server_thread = threading.Thread(
+            target=uvicorn.run,
+            kwargs={
+                "app": fastapi_app,
+                "host": "127.0.0.1",
+                "port": 8766,
+                "log_level": "warning",  # Only warnings and above
+            },
+        )
+        server_thread.daemon = True
+        server_thread.start()
+
+        print("Server thread started, waiting for it to initialize...")
+
+        # Wait for server to start
+        await asyncio.sleep(2)
+
+        # Create socketio client
+        sio = socketio.AsyncClient(logger=False, engineio_logger=False)
+
+        @sio.event
+        async def connect():
+            print("SocketIO client connected successfully")
+            results.connected = True
+            results.received_events.append("connect")
+
+        @sio.event
+        async def connect_error(data):
+            print(f"SocketIO connection error: {data}")
+            results.connection_errors.append(data)
+            results.received_events.append(f"connect_error: {data}")
+
+        @sio.event
+        async def disconnect():
+            print("SocketIO client disconnected")
+            results.connected = False
+            results.received_events.append("disconnect")
+
+        @sio.on("threshold")
+        async def on_threshold(data):
+            print(f"Received threshold event: {data}")
+            results.threshold = data
+            results.received_events.append(f"threshold: {data}")
+
+        @sio.on("project_rewards")
+        async def on_project_rewards(data):
+            print("Received project_rewards event")
+            results.project_rewards = data
+            results.received_events.append(f"project_rewards: {data}")
+
+        @sio.on("project_donors")
+        async def on_project_donors(data):
+            project = data.get("project")
+            donors = data.get("donors")
+            print(f"Received project_donors event for project: {project}")
+            results.project_donors[project] = donors
+            results.received_events.append(f"project_donors: {project}")
+
+        @sio.on("exception")
+        async def on_exception(data):
+            print(f"Received exception event: {data}")
+            results.exception = data
+            results.received_events.append(f"exception: {data}")
+
+        # Catch any other events
+        @sio.on("*")
+        async def catch_all(event, data):
+            print(f"Received unhandled event: {event}, data: {data}")
+            results.received_events.append(f"other: {event}")
```

Comment:
i haven't noticed this previously but why do all these definition sit here inside the `run_test` courutine? it should be defined somewhere else and the `results` doesn't have to be dependent on the closure, it may be passed as an argument